crates/cli/src/config.rs

```diff
@@ -511,15 +586,76 @@ impl Config {
         Self {
             home: RawConfig::new_with_localhost(),
             home_path,
+            doc: Default::default(),
         }
     }
 
+    /// Returns a preserving copy of [Config].
+    fn doc(&self) -> DocumentMut {
+        let mut doc = self.doc.clone();
+
+        let mut set_value = |key: &str, value: Option<&str>| {
+            set_opt_value(&mut doc, key, value);
+        };
+        // Intentionally use a destructuring assignment in case the fields change...
+        let RawConfig {
+            default_server,
+            server_configs,
+            web_session_token,
+            spacetimedb_token,
+        } = &self.home;
+
+        set_value(DEFAULT_SERVER_KEY, default_server.as_deref());
+        set_value(WEB_SESSION_TOKEN_KEY, web_session_token.as_deref());
+        set_value(SPACETIMEDB_TOKEN_KEY, spacetimedb_token.as_deref());
+
+        // Short-circuit if there are no servers.
+        if server_configs.is_empty() {
+            doc.remove(SERVER_CONFIGS_KEY);
+            return doc;
+        }
+        // ... or if there are no server_configs to edit.
+        let server_configs = if let Some(cfg) = doc.get_mut(SERVER_CONFIGS_KEY).and_then(Item::as_array_of_tables_mut) {
+            cfg
+        } else {
+            doc[SERVER_CONFIGS_KEY] = Item::ArrayOfTables(server_configs.iter().map(ServerConfig::as_table).collect());
+            return doc;
+        };
+
+        let mut new_configs = self
+            .home
+            .server_configs
+            .iter()
+            .map(|cfg| (cfg.nick_or_host(), cfg))
+            .collect::<HashMap<_, _>>();
+
+        // Update the existing servers.
```

Comment:
```suggestion
        // Update the existing servers, and remove deleted servers.
        // We'll add new servers later.
        // We do this somewhat elaborate dance rather than just overwriting the config
        // in order to preserve the order and formatting of pre-existing server configs in the file.
```